eth/downloader/queue.go

```diff
@@ -365,6 +458,77 @@ func (q *queue) Deliver(id string, blocks []*types.Block) (err error) {
 	return nil
 }
 
+// Deliver injects a block body retrieval response into the download queue.
+func (q *queue) Deliver(id string, transactions [][]*types.Transaction, uncles [][]*types.Header) (err error) {
+	q.lock.Lock()
+	defer q.lock.Unlock()
+
+	// Short circuit if the block bodies were never requested
+	request := q.pendPool[id]
+	if request == nil {
+		return errNoFetchesPending
+	}
+	delete(q.pendPool, id)
+
+	// If no block bodies were retrieved, mark them as unavailable for the origin peer
+	if len(transactions) == 0 || len(uncles) == 0 {
```

Comment:
Ah wait I see, it's slice of slice :) Could transactions and uncles here be renamed to make this more clear, e.g. txLists, uncleLists or some such.